File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
class Solution {
public:
    int mxDmtr = 0;

    int dfs(TreeNode* root) {
        if (!root) return 0;

        int lh = dfs(root->left);
        int rh = dfs(root->right);

        mxDmtr = max(mxDmtr, lh + rh);

        return 1 + max(lh, rh);
    }

    int diameterOfBinaryTree(TreeNode* root) {
        dfs(root);

        return mxDmtr;
    }
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
class Solution {
    public int height(TreeNode root, int arr[]){
        if(root == null) return 0;

        int lh = height(root.left, arr);
        int rh = height(root.right, arr);

        arr[0] = Math.max(arr[0], lh + rh);

        return Math.max(lh, rh) + 1;
    }

    public int diameterOfBinaryTree(TreeNode root) {
        if(root ==  null) return 0;
        int arr[] = new int[1];

        height(root, arr);

        return arr[0];
    }
}